OsmAndMapCreator/src/net/osmand/BinaryMerger.java

```diff
@@ -0,0 +1,329 @@
+			writer.startCityBlockIndex(type);
+			for (City city : cities) {
+				refs.add(writer.writeCityHeader(city, city.getType().ordinal(), tagRules));
+			}
+			for (int i = 0; i != refs.size(); i++) {
+				BinaryFileReference ref = refs.get(i);
+				City city = cities.get(i);
+				IndexAddressCreator.putNamedMapObject(namesIndex, city, ref.getStartPointer());
+				List<Street> streets = new ArrayList(city.getStreets());
+				Map<Street, List<Node>> streetNodes = new LinkedHashMap<Street, List<Node>>();
+				writer.writeCityIndex(city, streets, streetNodes, ref, tagRules);
+				// register postcodes and name index
+				for (Street s : streets) {
+					IndexAddressCreator.putNamedMapObject(namesIndex, s, s.getFileOffset());
+					for (Building b : s.getBuildings()) {
+						if (city.getPostcode() != null && b.getPostcode() == null) {
+							b.setPostcode(city.getPostcode());
+						}
+						if (b.getPostcode() != null) {
+							if (!postcodes.containsKey(b.getPostcode())) {
+								City p = City.createPostcode(b.getPostcode());
+								p.setLocation(b.getLocation().getLatitude(), b.getLocation().getLongitude());
+								postcodes.put(b.getPostcode(), p);
+							}
+							City post = postcodes.get(b.getPostcode());
+							Street newS = post.getStreetByName(s.getName());
+							if(newS == null) {
+								newS = new Street(post);
+								newS.copyNames(s);
+								newS.setLocation(s.getLocation().getLatitude(), s.getLocation().getLongitude());
+								newS.setId(s.getId());
+								post.registerStreet(newS);
+							}
+							newS.addBuildingCheckById(b);
+						}
+					}
+				}
+			}
+			writer.endCityBlockIndex();
+		}
+		writer.writeAddressNameIndex(namesIndex);
+		writer.endWriteAddressIndex();
+	}
+
+	@SuppressWarnings("unchecked")
+	public static List<Float> combineParts(File fileToExtract, Map<File, String> partsToExtractFrom) throws IOException {
+		BinaryMapIndexReader[] indexes = new BinaryMapIndexReader[partsToExtractFrom.size()];
+		RandomAccessFile[] rafs = new RandomAccessFile[partsToExtractFrom.size()];
+
+		LinkedHashSet<Float>[] partsSet = new LinkedHashSet[partsToExtractFrom.size()];
+		int c = 0;
+		Set<String> addressNames = new LinkedHashSet<String>();
+
+
+		int version = -1;
+		// Go through all files and validate conistency
+		for (File f : partsToExtractFrom.keySet()) {
+			if (f.getAbsolutePath().equals(fileToExtract.getAbsolutePath())) {
+				System.err.println("Error : Input file is equal to output file " + f.getAbsolutePath());
+				return null;
+			}
+			rafs[c] = new RandomAccessFile(f.getAbsolutePath(), "r");
+			indexes[c] = new BinaryMapIndexReader(rafs[c], f);
+			partsSet[c] = new LinkedHashSet<Float>();
+			if (version == -1) {
+				version = indexes[c].getVersion();
+			} else {
+				if (indexes[c].getVersion() != version) {
+					System.err.println("Error : Different input files has different input versions " + indexes[c].getVersion() + " != " + version);
+					return null;
+				}
+			}
+
```

Comment:
In general this code is not needed if we focus on just merging (and not deleting any part)